flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StateConfigUtil.java

```diff
@@ -46,19 +46,24 @@ public static StateTtlConfig createTtlConfig(long retentionTime) {
         }
     }
 
-    public static boolean isStateImmutableInStateBackend(KeyedStateBackend<?> stateBackend) {
+    public static boolean isStateImmutableInStateBackend(KeyedStateBackend<?> keyedStateBackend) {
         // TODO: remove the hard code check once FLINK-21027 is supported
         // state key and value is immutable only when using rocksdb state backend and timer
+        KeyedStateBackend<?> rootKeyedStateBackend =
+                keyedStateBackend instanceof DelegateKeyedStateBackend
+                        ? ((DelegateKeyedStateBackend<?>) keyedStateBackend)
+                                .getDelegatedKeyedStateBackend()
+                        : keyedStateBackend;
+
         boolean isRocksDbState =
-                ROCKSDB_KEYED_STATE_BACKEND.equals(stateBackend.getClass().getCanonicalName());
-        boolean isHeapTimer = false;
-        if (stateBackend instanceof AbstractKeyedStateBackend) {
-            // currently, requiresLegacySynchronousTimerSnapshots()
-            // indicates the underlying uses heap-bsased timer
-            isHeapTimer =
-                    ((AbstractKeyedStateBackend<?>) stateBackend)
-                            .requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT);
-        }
+                ROCKSDB_KEYED_STATE_BACKEND.equals(
+                        rootKeyedStateBackend.getClass().getCanonicalName());
```

Comment:
I think this `getDelegatedKeyedStateBackend` is not necessary anymore as well as pre-existing class check (is rocks db) after you added `requiresLegacySynchronousTimerSnapshots` to `KeyedStateBackend`.

So `isStateImmutableInStateBackend` can simply return `keyedStateBackend.requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT)`;

Reply:
keyedStateBackend.requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT) == true

means it  `use rocksdb` && `uses Heap Timer`

What needs to return is `use rocksdb` && `NOT use Heap Timer`

They are not the same, and you can not simply return `keyedStateBackend.requiresLegacySynchronousTimerSnapshots(CheckpointType.CHECKPOINT);`

Reply:
`requiresLegacySynchronousTimerSnapshots` is overriden in `RocksDBKeyedStateBackend` so there is no need to check the class.
Or am I missing something?

Reply:
The returned logic of isStateImmutableInStateBackend() is not the same as what requiresLegacySynchronousTimerSnapshots() returns, please check my highlights of the last comment.

requiresLegacySynchronousTimerSnapshots() == false only indicates it does not uses a heap timer; it does not indicate it is a rocksdb statebackend.

the required logic needs requiresLegacySynchronousTimerSnapshots() == false && uses rocks db

Reply:
Maybe it's easier to discuss having a code change.
I published a hotfix PR: https://github.com/apache/flink/pull/15046

Reply:
That logically makes more sense.